parts/linux/cloud-init/artifacts/cse_install.sh

```diff
@@ -140,15 +140,35 @@ installNetworkPlugin() {
     rm -rf $CNI_DOWNLOADS_DIR & 
 }
 
-
-downloadCredentalProvider() {
+# downloadCredentialProvider is most-likely called during build time by install-dependencies.sh. 
+# It can be called during node provisioning by cse_config.sh, meaning CREDENTIAL_PROVIDER_DOWNLOAD_URL is set by a passed in linuxCredentialProviderURL.
+downloadCredentialProvider() {
     mkdir -p $CREDENTIAL_PROVIDER_DOWNLOAD_DIR
-    CREDENTIAL_PROVIDER_TGZ_TMP=${CREDENTIAL_PROVIDER_DOWNLOAD_URL##*/} # Use bash builtin ## to remove all chars ("*") up to the final "/"
-    retrycmd_get_tarball 120 5 "$CREDENTIAL_PROVIDER_DOWNLOAD_DIR/$CREDENTIAL_PROVIDER_TGZ_TMP" "$CREDENTIAL_PROVIDER_DOWNLOAD_URL" || exit $ERR_CREDENTIAL_PROVIDER_DOWNLOAD_TIMEOUT
+
+    CREDENTIAL_PROVIDER_DOWNLOAD_URL="${CREDENTIAL_PROVIDER_DOWNLOAD_URL:=}"
+    if [[ -n "${CREDENTIAL_PROVIDER_DOWNLOAD_URL}" ]]; then
+        # CREDENTIAL_PROVIDER_DOWNLOAD_URL is set by linuxCredentialProviderURL
+        # The version in the URL is unknown. An acs-mirror or registry URL could be passed meaning the version must be extracted from the URL. 
+        CREDENTIAL_PROVIDER_VERSION=$(echo "$CREDENTIAL_PROVIDER_DOWNLOAD_URL" | grep -oP 'v\d+(\.\d+)*' | sed 's/^v//' | head -n 1)
+    else
+        CREDENTIAL_PROVIDER_DOWNLOAD_URL="https://acs-mirror.azureedge.net/cloud-provider-azure/v${CREDENTIAL_PROVIDER_VERSION}/binaries/azure-acr-credential-provider-linux-${CPU_ARCH}-v${CREDENTIAL_PROVIDER_VERSION}.tar.gz"
```

Comment:
not sure why we need to redefine DOWNLOAD_URL here. Is it trying to make sure the default value?
If so seems still possible miss the variable assignment of CREDENTIAL_PROVIDER_VERSION

Reply:
This is because in the if statement `CREDENTIAL_PROVIDER_DOWNLOAD_URL` is set by `linuxCredentialProviderURL` and I want to make sure I don't override that. But, there is a chance that CREDENTIAL_PROVIDER_DOWNLOAD_URL` is set by `linuxCredentialProviderURL` and a container registry is provided, in that case I want to take the version that is in the URL string, and use it in the string registry URL

Otherwise, I changed how the function is called, `CREDENTIAL_PROVIDER_DOWNLOAD_URL` is no longer passed in so it needs to be defaulted in the else.